posthog/management/commands/backfill_distinct_id_overrides.py

```diff
@@ -0,0 +1,125 @@
+
+    @property
+    def size(self):
+        return self.upper - self.lower
+
+    def split(self) -> Iterator[Range]:
+        if self.size < 2:
+            raise ValueError("cannot split range")
+
+        midpoint = self.lower + (self.upper - self.lower) // 2
+        return iter(
+            [
+                Range(self.lower, midpoint),
+                Range(midpoint, self.upper),
+            ]
+        )
+
+
+@dataclass
+class BackfillQuery:
+    team_id: int
+    range: Range = Range(0, 2**64)
+
+    def split(self) -> Iterator[BackfillQuery]:
+        for chunk in self.range.split():
+            yield BackfillQuery(self.team_id, chunk)
+
+    def execute(self, dry_run: bool = False) -> None:
+        query = """
+            SELECT
+                team_id,
+                distinct_id,
+                pdi.person_id as person_id,
+                -1 as version  -- overrides that come in via Kafka will overwrite this
+            FROM events
+            LEFT JOIN (
+                SELECT
+                    distinct_id,
+                    argMax(person_id, version) as person_id
+                FROM person_distinct_id2
+                WHERE
+                    team_id = %(team_id)s
+                    AND %(range_lower)s <= cityHash64(distinct_id)
+                    AND cityHash64(distinct_id) < %(range_upper)s
+                GROUP BY ALL
+            ) pdi
+                ON pdi.distinct_id = events.distinct_id
+            WHERE
+                team_id = %(team_id)s
+                and %(range_lower)s <= cityHash64(distinct_id)
+                and cityHash64(distinct_id) < %(range_upper)s
+                and events.person_id != pdi.person_id
+            GROUP BY ALL
+        """
+
+        parameters = {
+            "team_id": self.team_id,
+            "range_lower": self.range.lower,
+            "range_upper": self.range.upper,
+        }
+
+        if dry_run:
+            [(count,)] = sync_execute(f"SELECT count() FROM ({query})", parameters)
+            logger.info("%r would have inserted %r records.", self, count)
+        else:
+            # XXX: Nothing useful to report here, unfortunately... all that is
+            # returned is an empty result set.
+            sync_execute(
+                f"""
+                    INSERT INTO person_distinct_id_overrides
+                    (team_id, distinct_id, person_id, version)
+                    {query}
+                """,
+                parameters,
+            )
+
+
+def execute_backfill(query: BackfillQuery, dry_run: bool = False) -> None:
+    logger.info(f"Executing %r...", query)
+    try:
+        query.execute(dry_run=dry_run)
+    except ServerException as e:
+        if e.code not in {ErrorCodes.TOO_SLOW, ErrorCodes.TOO_MANY_ROWS}:
+            raise e
+        logger.warn(f"Caught %s when running %r! Trying smaller ranges...", e, query)
+        for chunk in query.split():
+            execute_backfill(chunk)
```

Comment:
There's probably a kinder and more intelligent way to do this than just recursively dividing the keyspace into smaller chunks until it doesn't error, but this seems good enough for now?

Reply:
After some ad hoc testing, I think it might be more effective to try and set a target size for the right hand side table, and split into how ever many chunks are needed to achieve that target size by using the value of `count() / target_size` with the presumption that `cityHash64(distinct_id)` is uniformly distributed. I'm not sure what that target size should be though: figuring that out might require some numberwanging.

As implemented, this has the drawback that queries can either be too slow because the right hand side table is too big (the join is too complex) or because the left hand side table requires scanning too many rows. Taking care of the size of the left hand side likely means breaking down the job by partitions when the size is unwieldily large.

Going to put this back as a draft while I look into doing that.